src/operator/tensor/diag_op-inl.h

```diff
@@ -104,42 +143,132 @@ inline bool DiagOpType(const nnvm::NodeAttrs& attrs,
   }
 };
 
-template<int req>
+template<int req, bool back>
 struct diag_gen {
   template<typename DType>
-  MSHADOW_XINLINE static void Map(int i, DType* out, const DType* a,
+  MSHADOW_XINLINE static void Map(index_t i, DType* out, const DType* a,
                                   mshadow::Shape<2> oshape, int k) {
     using namespace mxnet_op;
 
     auto j = unravel(i, oshape);
     if (j[1] == (j[0] + k)) {
       auto l = j[0] < j[1] ? j[0] : j[1];
-      KERNEL_ASSIGN(out[i], req, a[l]);
-    } else {
+      if (back) {
+        KERNEL_ASSIGN(out[l], req, a[i]);
+      } else {
+        KERNEL_ASSIGN(out[i], req, a[l]);
+      }
+    } else if (!back) {
       KERNEL_ASSIGN(out[i], req, static_cast<DType>(0));
     }
   }
 };
 
+template<typename xpu, bool back>
+void DiagOpProcess(const TBlob& in_data,
+                   const TBlob& out_data,
+                   const TShape& ishape,
+                   const TShape& oshape,
+                   index_t dsize,
+                   const DiagParam& param,
+                   mxnet_op::Stream<xpu> *s,
+                   const std::vector<OpReqType>& req) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  if (ishape.ndim() > 1) {
+    // input : (leading + i, body + i, trailing)
+    int x1 = CheckAxis(param.axis1.value(), ishape.ndim());
+    int x2 = CheckAxis(param.axis2.value(), ishape.ndim());
+
+    int idim = ishape.ndim(), odim = oshape.ndim();
+
+    int minx = x1, maxx = x2;
+    if (minx > maxx) {
+      std::swap(minx, maxx);
+    }
+
+    index_t oleading = 1,
+           obody = 1,
+           otrailing = 1;
+
+    for (int i = 0; i < minx; ++i) {
+      oleading *= ishape[i];
+    }
+    for (int i = minx + 1; i < maxx; ++i) {
+      obody *= ishape[i];
+    }
+    for (int i = maxx + 1; i < idim; ++i) {
+      otrailing *= ishape[i];
+    }
+
+    index_t ileading = oleading,
+        ibody = obody * ishape[minx],
+        itrailing = otrailing * ishape[maxx];
+
+    index_t stride1 = itrailing * obody,
+        stride2 = otrailing;
+
+    if (x1 == maxx) {
+      std::swap(stride1, stride2);
+    }
+    index_t offset;
+    int k = param.k.value();
+    if (k > 0) {
+      offset = stride2 * k;
+    } else if (k < 0) {
+      offset = stride1 * -k;
+    } else {
+      offset = 0;
+    }
+
+    MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+      MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
+        if (back && req[0] != kAddTo && req[0] != kNullOp) {
+          out_data.FlatTo1D<xpu, DType>(s) = 0;
+        }
+        if (ileading == 1) {
+          Kernel<diag<2, req_type, back>, xpu>::Launch(s, dsize, out_data.dptr<DType>(),
+                              in_data.dptr<DType>(), Shape2(obody, otrailing),
+                              Shape2(ibody, itrailing),
+                              stride1 + stride2, offset, oshape[odim - 1]);
+        } else {
+          Kernel<diag<3, req_type, back>, xpu>::Launch(s, dsize, out_data.dptr<DType>(),
```

Comment:
Why is the ndim template argument now fixed to 2 or 3 here? shouldnt it be generic using the odim variable you define in line 202?

Reply:
That is because for efficiency here I have merged some contiguous axes together, and the resulting shape cannot have more than 3 axes. In the original shape, there is no need to distinguish two axes if they are not separated by `axis1` or `axis2`.

Reply:
More efficiency is always great. But im having trouble following the abstraction to 2 or 3 axes. Can you elaborate some more on the approach/implementation?

Reply:
Okay. Let's assume that the input shape is `(s1, s2, ..., sa - 1, sa, sa+1, ... , sb - 1, sb, sb + 1, ... sn)` where `a = axis1, b = axis2`. What would be the output shape? It would be `(s1, s2, ..., sa - 1, sa + 1, ..., sb - 1, sb + 1, ... sn, sd)` where `sd` is the size of the diagonal of a `sa x sb` matrix. Note that `s1, s2, ..., sa - 1`, `sa + 1, sa + 2, ..., sb - 1` and `sb + 1, sb + 2, ... sn` remain untouched (directly mapped), and for this reason we can merge them respectively. To remove `sa` and `sb` we can also merge them into segments `sa + 1, sa + 2, ..., sb - 1` and `sb + 1, sb + 2 , ... , sn` respectively, thus reshaping the original input into `(s1 * s2 * ... * sa - 1, sa * sa + 1 * ... * sb - 1,  sb * sb + 1 * ... * sn)`, and the corresponding output into `(s1 * s2 * ... * sa - 1, sa + 1 * sa + 2 * ... * sb - 1, sb + 1 * sb + 2 * ... * sn, sd)`. In this way, to get `output(x1, x2, x3, x4)`, we can simply find the index of `input(x1, x2, x3)` plus an offset `x4 * (stride1 + stride2)` where `stride1` and `stride2` are strides of `axis1` and `axis2` in the input respectively.

Reply:
Ok cool that makes sense. Thats a nice simplification to compute the index mapping between input/output coordinates. It might be worth while to capture some of this optimization in comments in the code somewhere.

Reply:
@samskalicky I will take your advice! Thanks!